bookwyrm/tests/views/preferences/test_two_factor_auth.py

```diff
@@ -0,0 +1,195 @@
+        form = forms.ConfirmPasswordForm()
+        form.data["password"] = "password"
+        request = self.factory.post("", form.data)
+        request.user = self.local_user
+
+        with patch("bookwyrm.views.preferences.two_factor_auth.Edit2FA"):
+            result = view(request)
+        self.assertIsInstance(result, TemplateResponse)
+        self.assertEqual(result.status_code, 200)
+
+    def test_post_confirm_2fa(self, *_):
+        """check 2FA login works"""
+        view = views.Confirm2FA.as_view()
+        form = forms.Confirm2FAForm()
+        totp = pyotp.TOTP("UEWMVJHO23G5XLMVSOCL6TNTSSACJH2X")
+        form.data["otp"] = totp.now()
+        request = self.factory.post("", form.data)
+        request.user = self.local_user
+
+        with patch("bookwyrm.views.preferences.two_factor_auth.Confirm2FA"):
+            result = view(request)
+        self.assertIsInstance(result, TemplateResponse)
+        self.assertEqual(result.status_code, 200)
+
+    def test_get_disable_2fa(self, *_):
+        """there are so many views, this just makes sure it LOADS"""
+        view = views.Disable2FA.as_view()
+        request = self.factory.get("")
+        request.user = self.local_user
+        result = view(request)
+        self.assertIsInstance(result, TemplateResponse)
+        self.assertEqual(result.status_code, 200)
+
+    def test_post_disable_2fa(self, *_):
+        """check 2FA login works"""
+        view = views.Disable2FA.as_view()
+        request = self.factory.post("")
+        request.user = self.local_user
+
+        with patch("bookwyrm.views.preferences.two_factor_auth.Disable2FA"):
+            result = view(request)
+        self.assertIsInstance(result, TemplateResponse)
+        self.assertEqual(result.status_code, 200)
+
+    def test_get_login_with_2fa(self, *_):
+        """there are so many views, this just makes sure it LOADS"""
+        view = views.LoginWith2FA.as_view()
+        request = self.factory.get("")
+        request.user = self.local_user
+        result = view(request)
+        self.assertIsInstance(result, TemplateResponse)
+        self.assertEqual(result.status_code, 200)
+
+    # def test_post_login_with_2fa(self, *_):
+    #     """check 2FA login works"""
+    # NOTE this throws a redis error.
+    # Possibly because the function it tests wants to check the user database?
+    # can we mock that?
+    # view = views.LoginWith2FA.as_view()
+    # form = forms.Confirm2FAForm()
+    # totp = pyotp.TOTP('UEWMVJHO23G5XLMVSOCL6TNTSSACJH2X')
+
+    # form.data["otp"] = totp.now()
+    # request = self.factory.post("", form.data)
+    # request.user = self.local_user
+
+    # middleware = SessionMiddleware(request)
+    # middleware.process_request(request)
+    # request.session['2fa_auth_time'] = time.time()
+    # request.session['2fa_user'] = self.local_user.username
+    # request.session.save()
+
+    # with patch("bookwyrm.views.preferences.two_factor_auth.LoginWith2FA"):
```

Comment:
If you also add `patch("bookwyrm.views.preferences.two_factor_auth.login")`  here, you should be good to go. The response was a 302, though, not a TemplateResponse so you'll have to change the assertions.